pkg/kvstore/etcd.go

```diff
@@ -0,0 +1,372 @@
+		if err != nil {
+			return err
+		}
+
+		value, err := e.GetValue(keyPath)
+		if err != nil {
+			locker.Unlock()
+			return err
+		}
+		if value == nil {
+			return setID2Label(locker)
+		}
+		var consulLabels types.SecCtxLabel
+		if err := json.Unmarshal(value, &consulLabels); err != nil {
+			locker.Unlock()
+			return err
+		}
+		if consulLabels.RefCount() == 0 {
+			log.Infof("Recycling ID %d", baseID)
+			return setID2Label(locker)
+		}
+		locker.Unlock()
+
+		baseID++
+		if baseID > common.MaxSetOfLabels {
+			baseID = common.FirstFreeLabelID
+		}
+		if beginning == baseID {
+			return fmt.Errorf("reached maximum set of labels available.")
+		}
+	}
+}
+
+func (e *EtcdClient) updateSecLabelIDRef(secCtxLabels types.SecCtxLabel) error {
+	key := path.Join(common.LabelIDKeyPath, strconv.FormatUint(uint64(secCtxLabels.ID), 10))
+	return e.SetValue(key, secCtxLabels)
+}
+
+func (e *EtcdClient) setMaxLabelID(maxID uint32) error {
+	return e.SetMaxID(common.LastFreeLabelIDKeyPath, common.FirstFreeLabelID, maxID)
+}
+
+func (e *EtcdClient) setMaxServiceL4ID(maxID uint32) error {
+	return e.SetMaxID(common.LastFreeServiceIDKeyPath, common.FirstFreeServiceID, maxID)
+}
+
+func (e *EtcdClient) GASNewServiceL4ID(basePath string, baseID uint32, sl4 *types.ServiceL4ID) error {
+	setID2ServiceL4 := func(lockPair KVLocker) error {
+		defer lockPair.Unlock()
+		sl4.ServiceID = types.ServiceID(baseID)
+		keyPath := path.Join(basePath, strconv.FormatUint(uint64(sl4.ServiceID), 10))
+		if err := e.SetValue(keyPath, sl4); err != nil {
+			return err
+		}
+		return e.setMaxServiceL4ID(baseID + 1)
+	}
+
+	beginning := baseID
+	for {
+		log.Debugf("Trying to acquire a new free ID %d", baseID)
+		keyPath := path.Join(basePath, strconv.FormatUint(uint64(baseID), 10))
+
+		locker, err := e.LockPath(GetLockPath(keyPath))
+		if err != nil {
+			locker.Unlock()
+			return err
+		}
+
+		value, err := e.GetValue(keyPath)
+		if err != nil {
+			locker.Unlock()
+			return err
+		}
+		if value == nil {
+			return setID2ServiceL4(locker)
+		}
+		var consulServiceL4ID types.ServiceL4ID
+		if err := json.Unmarshal(value, &consulServiceL4ID); err != nil {
+			locker.Unlock()
+			return err
+		}
+		if consulServiceL4ID.ServiceID == 0 {
+			log.Infof("Recycling Service ID %d", baseID)
+			return setID2ServiceL4(locker)
+		}
+		locker.Unlock()
+
+		baseID++
+		if baseID > common.MaxSetOfServiceID {
+			baseID = common.FirstFreeServiceID
+		}
+		if beginning == baseID {
+			return fmt.Errorf("reached maximum set of serviceIDs available.")
+		}
+	}
+}
+
+func (e *EtcdClient) DeleteTree(path string) error {
+	_, err := e.Delete(ctx.Background(), path, client.WithPrefix())
+	return err
+}
+
+// GetWatcher watches for kvstore changes in the given key. Triggers the returned channel
+// every time the key path is changed.
+func (e *EtcdClient) GetWatcher(key string, timeSleep time.Duration) <-chan []uint32 {
+	ch := make(chan []uint32, 100)
```

Comment:
Let's keep it as is for now.

This function is still WIP because we still haven't figure it out how to take advantage of this. The purpose of this watcher is to get the `security IDs` that were changed, passing the pointer to the caller. That caller will recalculate the policy tree only for the `security IDs` that were changed. Right now we are recalculating all policy trees whenever something is changed.